Mathlib/Algebra/Module/LocalizedModule.lean

```diff
@@ -1079,6 +1081,71 @@ theorem mkOfAlgebra {R S S' : Type _} [CommRing R] [CommRing S] [CommRing S'] [A
 
 end Algebra
 
+variable {A : Type _}
+  [CommRing A] [Algebra R A] [Module A M'] [IsScalarTower R A M'] [IsLocalization S A]
+
+
+/-- If `(f : M →ₗ[R] M')` is a localization of modules, then the map
+`(localization S) × M → N, (s, m) ↦ s • f m` is the tensor product (insomuch as it is the universal
+bilinear map).
+In particular, there is an isomorphism between `LocalizedModule S M` and `(Localization S) ⊗[R] M`
+given by `m/s ↦ (1/s) ⊗ₜ m`.
+-/
+theorem isBaseChange : IsBaseChange A f := by
+  refine' IsBaseChange.of_lift_unique _ (fun Q _ _ _ _ g ↦ _)
+  have := IsLocalizedModule.is_universal S f g <| by
+    intro s
+    simp_rw [Module.End_isUnit_iff, Function.bijective_iff_existsUnique,
+      Module.algebraMap_end_apply]
+    intro q
+    refine' ⟨(IsLocalization.mk' _ 1 s : A) • q, _, _⟩
+    · simp only [← smul_assoc, IsLocalization.smul_mk'_self, map_one, one_smul]
+    · rintro q rfl
+      simp only [smul_comm _ (s : R), ← smul_assoc, IsLocalization.smul_mk'_self, map_one, one_smul]
+  rcases this with ⟨ℓ, h₁, h₂⟩
+  -- Should this be refactored into a `linear_map.extend_scalars` lemma? If so, what would the
+  -- predicate have to be? Just `map_smul'`?
+  let g' : M' →ₗ[A] Q :=
+    { toFun := ℓ.toFun
+      map_add' := ℓ.map_add'
+      map_smul' := by
+        intros r x
+        rw [RingHom.id_apply]
+        rcases IsLocalization.mk'_surjective S r with ⟨r,s,rfl⟩
+        conv_lhs =>
+          dsimp
+          rw [← one_smul A (ℓ _), ← IsLocalization.mk'_self' _ (x := s),  ← mul_one r,
+            ← mul_one (↑s : R)]
+          conv =>
+            -- 'repeat' appears to discharge goals
+            -- https://github.com/leanprover/lean4/pull/2357
+            repeat rw [← IsLocalization.smul_mk' ]
+          rw [smul_assoc r, ℓ.map_smul]
+          conv in (↑s • _) • _ =>
+            rw [smul_comm]
+          conv in (↑s • _) • _ =>
+            rw [smul_assoc, smul_comm]
+          dsimp
+          rw [← ℓ.map_smul (↑s : ↑ S)]
+          conv in (occs := 1) _ • _ • _ =>
+            rw [← smul_assoc]
+          arg 2; rw [← smul_assoc]
+        iterate 2 rw [IsLocalization.smul_mk', mul_one]
+        rw [IsLocalization.mk'_self, one_smul]; rfl }
+  use g'
+  have g'_extends_scalars : LinearMap.restrictScalars R g' = ℓ := by
+    ext
+    simp [LinearMap.restrictScalars_apply, LinearMap.toFun_eq_coe]
+  constructor
+  · dsimp only
+    rwa [g'_extends_scalars]
+  · rintro g'' h
+    have := h₂ (LinearMap.restrictScalars R g'') h
+    rw [← g'_extends_scalars] at this
+    ext x
+    apply_fun fun f => f x at this
+    simpa [LinearMap.restrictScalars_apply]
```

Comment:
If we add the following definition somewhere (maybe to `RingTheory/Localization/Module.lean`):
```lean
section move_me

variable (A)
variable {M N : Type _}
  [AddCommMonoid M] [Module R M] [Module A M] [IsScalarTower R A M]
  [AddCommMonoid N] [Module R N] [Module A N] [IsScalarTower R A N]

/-- An `R`-linear map between two `S⁻¹R`-modules is actually `S⁻¹R`-linear. -/
def _root_.LinearMap.extendScalarsOfIsLocalization (f : M →ₗ[R] N) : M →ₗ[A] N where
  toFun := f
  map_add' := f.map_add
  map_smul' := by
    intro r m
    simp only [RingHom.id_apply]
    rcases IsLocalization.mk'_surjective S r with ⟨r, s, rfl⟩
    conv_lhs =>
      rw [← one_smul A (f _), ← IsLocalization.mk'_self' _ (x := s),  ← mul_one r,
        ← mul_one (↑s : R)]
      conv =>
        -- 'repeat' appears to discharge goals
        -- https://github.com/leanprover/lean4/pull/2357
        repeat rw [← IsLocalization.smul_mk' ]
      rw [smul_assoc r, f.map_smul]
      conv in (↑s • _) • _ =>
        rw [smul_comm]
      conv in (↑s • _) • _ =>
        rw [smul_assoc, smul_comm]
      rw [← f.map_smul (↑s : ↑ S)]
      conv in (occs := 1) _ • _ • _ =>
        rw [← smul_assoc]
      arg 2; rw [← smul_assoc]
    iterate 2 rw [IsLocalization.smul_mk', mul_one]
    rw [IsLocalization.mk'_self, one_smul]

@[simp] lemma _root_.LinearMap.restrictScalars_extendScalarsOfIsLocalization (f : M →ₗ[R] N) :
    (f.extendScalarsOfIsLocalization S A).restrictScalars R = f := rfl

end move_me
```
then this proof can become:
```suggestion
  rcases this with ⟨ℓ, rfl, h₂⟩
  refine' ⟨ℓ.extendScalarsOfIsLocalization S A, by simp, fun g'' h ↦ _⟩
  have := h₂ (LinearMap.restrictScalars R g'') h
  rw [← ℓ.restrictScalars_extendScalarsOfIsLocalization S A] at this
  ext x
  apply_fun fun f => f x at this
  simpa [LinearMap.restrictScalars_apply]
```

Reply:
thanks! this is the refactor I was vaguely thinking about :laughing: